src/event/net/cares/Channel.cxx

```diff
@@ -42,36 +43,41 @@ class Channel::Socket {
   }
 };
 
+void Channel::sock_state_cb(void *data, ares_socket_t socket_fd, int readable, int writable)
+{
+  unsigned events = 0;
+    if (readable) events |= SocketEvent::READ;
+    if (writable) events |= SocketEvent::WRITE;
+
+    if (events != 0){
+      auto &channel= *static_cast<Channel*>(data);
+      channel.sockets.emplace_front(channel, SocketDescriptor{socket_fd}, events);
+    }
+}
+
 Channel::Channel(EventLoop &event_loop)
     : defer_update_sockets(event_loop, BIND_THIS_METHOD(UpdateSockets)),
       timeout_event(event_loop, BIND_THIS_METHOD(OnTimeout))
 {
-  int code = ares_init(&channel);
+  struct ares_options options;
+
+  options.sock_state_cb = & Channel::sock_state_cb;
+  options.sock_state_cb_data = this;
+
+  int optmask = 0;
+  optmask |= ARES_OPT_SOCK_STATE_CB;
+
+  int code = ares_init_options(&channel,&options,optmask);
   if (code != 0) throw Error(code, "ares_init() failed");
 }
 
 Channel::~Channel() noexcept { ares_destroy(channel); }
 
+
 void
 Channel::UpdateSockets() noexcept
 {
   timeout_event.Cancel();
-  sockets.clear();
```

Comment:
you are removing this. We need  a way to clear the sockets, otherwise we have a memory leak.